File: leetcode/cpp/all-paths-from-source-to-target.cpp
/* -------------------------------------------------------------------------- */
/*        https://leetcode.com/problems/all-paths-from-source-to-target       */
/* -------------------------------------------------------------------------- */
#include <stack>
#include <vector>
using namespace std;

class Solver {
   private:
    vector<vector<int>>& graph;

   public:
    Solver(vector<vector<int>>& graph) : graph(graph) {}

    // Beats 7.2%
    // 60ms, 30.2MB
    vector<vector<int>> solve_recursive(int v) {
        if (v == graph.size() - 1) {
            return {{v}};
        }

        vector<vector<int>> paths_from_v;
        for (const auto nbr : graph[v]) {
            const auto nbr_path = solve_recursive(nbr);
            paths_from_v.insert(std::end(paths_from_v), std::begin(nbr_path), std::end(nbr_path));
        }

        return std::move(paths_from_v);
    }

    // Beats 67.67%
    // 10ms, 14.8 MB
    vector<vector<int>> solve_iterative() {
        int target = graph.size() - 1;

        vector<vector<int>> paths;
        stack<vector<int>> st;

        st.push({0});
        while (!st.empty()) {
            auto path = std::move(st.top());
            st.pop();
            for (const auto& nbr : graph[path.back()]) {
                auto path_copy = path;
                path_copy.push_back(nbr);

                if (nbr == target) {
                    paths.push_back(std::move(path_copy));
                } else {
                    st.push(std::move(path_copy));
                }
            }
        }

        return std::move(paths);
    }
};

vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) { return Solver(graph).solve_iterative(); }

File: leetcode/rust/src/all_paths_from_source_to_target.rs
/* -------------------------------------------------------------------------- */
/*        https://leetcode.com/problems/all-paths-from-source-to-target       */
/* -------------------------------------------------------------------------- */
pub struct Solver {
    graph: Vec<Vec<i32>>,
}

impl Solver {
    pub fn new(graph: Vec<Vec<i32>>) -> Self {
        Solver { graph }
    }

    // 10ms, 14.8 MB
    // Beats 15%
    pub fn solve_recursive(&self, v: i32) -> Vec<Vec<i32>> {
        if v as usize == self.graph.len() - 1 {
            vec![vec![v]]
        } else {
            let mut paths_from_v: Vec<Vec<i32>> = Vec::new();
            for nbr in self.graph[v as usize].iter() {
                paths_from_v.append(&mut self.solve_recursive(*nbr));
            }

            for path in paths_from_v.iter_mut() {
                path.insert(0, v);
            }

            paths_from_v
        }
    }

    // 2ms, 3MB
    // Beats 100%
    pub fn solver_iterative(&self) -> Vec<Vec<i32>> {
        let target = self.graph.len() - 1;

        let mut paths: Vec<Vec<i32>> = Vec::new();
        let mut stack: Vec<Vec<i32>> = Vec::new();
        stack.push(vec![0]);

        while let Some(top_path) = stack.pop() {
            let last_node = top_path.last().unwrap();
            for nbr in self.graph[*last_node as usize].iter() {
                let mut path = top_path.clone();
                path.push(*nbr);

                if *nbr as usize == target {
                    paths.push(path)
                } else {
                    stack.push(path)
                }
            }
        }

        paths
    }
}
